File: coordinator.js
/* coordinator.js — weekbeeld voor een coördinator met een eigen lijst leerlingen.

   Andere vraag dan bij de teamleider: niet "wie springt eruit in een afdeling",
   maar "hoe ging deze week bij mijn twintig". Daarom staat de huidige week
   uitgeschreven per lesdag, met de weken ervoor als streepjes erachter. */
(function () {
  'use strict';

  var D          = window.DATA || {};
  var LEERLINGEN = D.leerlingen || [];
  var CODES      = D.codes || {};
  var WEKEN      = D.weken || [];
  var STARTS     = D.weekStarts || [];
  var HUIDIG     = WEKEN.length - 1;          // de laatste week is deze week

  var DAGNAAM = ['ma', 'di', 'wo', 'do', 'vr'];

  function codeNaam(c) { return (CODES[c] && CODES[c].naam) || c; }
  function esc(s) {
    return String(s == null ? '' : s).replace(/[&<>"]/g, function (c) {
      return ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' })[c];
    });
  }
  function datumVan(wi, d) {
    var start = STARTS[wi];
    if (!start) return '';
    var dt = new Date(start + 'T00:00:00');
    dt.setDate(dt.getDate() + d);
    var p = function (n) { return String(n).padStart(2, '0'); };
    return dt.getFullYear() + '-' + p(dt.getMonth() + 1) + '-' + p(dt.getDate());
  }
  function datumKort(iso) {
    if (!iso) return '';
    var d = new Date(iso + 'T00:00:00');
    if (isNaN(d)) return iso;
    var mnd = ['jan','feb','mrt','apr','mei','jun','jul','aug','sep','okt','nov','dec'];
    return d.getDate() + ' ' + mnd[d.getMonth()];
  }

  /* Registraties van één leerling, gegroepeerd op datum. */
  function perDatum(l) {
    var uit = {};
    (l.entries || []).forEach(function (e) { (uit[e.datum] = uit[e.datum] || []).push(e); });
    return uit;
  }
  function tel(entries, soort) {
    return entries.filter(function (e) { return e.soort === soort; }).length;
  }
  function vanWeek(l, wi) {
    return (l.entries || []).filter(function (e) { return e.week === wi; });
  }

  /* ── Kop met de cijfers van deze week ───────────────────────────────────── */
  function tekenKop() {
    var deze = [];
    LEERLINGEN.forEach(function (l) { deze = deze.concat(vanWeek(l, HUIDIG)); });
    var raak = LEERLINGEN.filter(function (l) { return vanWeek(l, HUIDIG).length > 0; }).length;

    var tegels = [
      { k: 'Leerlingen', v: LEERLINGEN.length, d: raak + ' met iets deze week' },
      { k: 'Ongeoorloofd', v: tel(deze, 'ong'), u: 'u', d: 'deze week' },
      { k: 'Te laat', v: tel(deze, 'laat'), d: 'deze week' },
      { k: 'Geoorloofd', v: tel(deze, 'geo'), u: 'u', d: 'ziek, arts, verlof' }
    ];
    document.getElementById('tiles').innerHTML = tegels.map(function (t) {
      return '<div class="tile"><div class="k">' + t.k + '</div>' +
        '<div class="v">' + t.v + (t.u ? '<small>' + t.u + '</small>' : '') + '</div>' +
        '<div class="d">' + t.d + '</div></div>';
    }).join('');

    var w = WEKEN[HUIDIG];
    if (w) {
      document.getElementById('weeklabel').textContent = w.label + ' · ' + w.sub;
    }
  }

  /* ── De kaarten ─────────────────────────────────────────────────────────── */
  function dagCel(l, dm, wi, d) {
    var iso = datumVan(wi, d);
    var es = (dm[iso] || []).slice().sort(function (a, b) { return (a.uur || 0) - (b.uur || 0); });
    var zwaarte = es.some(function (e) { return e.soort === 'ong'; }) ? 'ong'
                : es.some(function (e) { return e.soort === 'laat'; }) ? 'laat'
                : es.length ? 'geo' : 'niks';
    var chips = es.map(function (e) {
      return '<span class="wchip ' + e.soort + '" data-tip="' +
        esc(codeNaam(e.code) + (e.vak ? ' · ' + e.vak : '') + ' · ' + e.uurlabel) + '">' +
        esc(e.code) + '</span>';
    }).join('');
    return '<div class="wdag ' + zwaarte + '">' +
      '<span class="wdag-kop">' + DAGNAAM[d] + '</span>' +
      '<span class="wdag-body">' + (chips || '<span class="wleeg">·</span>') + '</span>' +
    '</div>';
  }

  function streepjes(l) {
    var dm = perDatum(l);
    var uit = '';
    for (var wi = 0; wi < HUIDIG; wi++) {
      var dagen = '';
      for (var d = 0; d < 5; d++) {
        var es = dm[datumVan(wi, d)] || [];
        var soort = es.some(function (e) { return e.soort === 'ong'; }) ? 'ong'
                  : es.some(function (e) { return e.soort === 'laat'; }) ? 'laat'
                  : es.length ? 'geo' : 'niks';
        var tip = es.length ? (es[0].daglabel + ' · ' + es.map(function (e) { return e.code; }).join(' ')) : '';
        dagen += '<i class="pdag ' + soort + (es.length > 2 ? ' veel' : '') + '"' +
                 (tip ? ' data-tip="' + esc(tip) + '"' : '') + '></i>';
      }
      uit += '<span class="pweek" data-tip="' + esc(WEKEN[wi].label + ' · ' + WEKEN[wi].sub) +
             '">' + dagen + '</span>';
    }
    return uit;
  }

  function logboekBlok(l) {
    var items = l.logboek || [];
    if (!D.logboekOpgehaald) return '';
    if (!items.length) return '<div class="lb-leeg">Nog geen logboek dit schooljaar</div>';
    var rijen = items.map(function (item) {
      var kop = esc(datumKort(item.datum) || item.datum) + ' · ' + esc(item.titel) +
                (item.door ? ' · ' + esc(item.door) : '');
      return '<details class="lb-item"><summary>' + kop + '</summary>' +
        (item.tekst ? '<pre class="lb-tekst">' + esc(item.tekst) + '</pre>'
                    : '<p class="ct-leeg">Tekst niet meegenomen in dit bestand.</p>') +
        '</details>';
    }).join('');
    return '<details class="logboek"><summary class="lb-kop">Logboek (' + items.length + ')' +
      (l.logboekDitJaar ? '' : ' — niets van dit schooljaar') + '</summary>' + rijen + '</details>';
  }

  function tekenKaarten() {
    /* Wie deze week het meest had, staat bovenaan. */
    var lijst = LEERLINGEN.slice().sort(function (a, b) {
      var wa = vanWeek(a, HUIDIG), wb = vanWeek(b, HUIDIG);
      var sa = tel(wa, 'ong') * 3 + tel(wa, 'laat') * 2 + tel(wa, 'geo');
      var sb = tel(wb, 'ong') * 3 + tel(wb, 'laat') * 2 + tel(wb, 'geo');
      return sb - sa || a.naam.localeCompare(b.naam, 'nl');
    });

    document.getElementById('kaarten').innerHTML = lijst.map(function (l) {
      var dm = dagenVan(l);
      var deze = vanWeek(l, HUIDIG);
      var rustig = deze.length === 0;
      var sub = l.klas + (l.mentorgroep ? ' · ' + l.mentorgroep : '') +
                (l.mentor ? ' · mentor ' + l.mentor : '');
      var cijfers = rustig
        ? '<span class="wrust">niets deze week</span>'
        : '<span class="wcijfers">' +
          (tel(deze, 'ong') ? '<b>' + tel(deze, 'ong') + ' u</b> ongeoorloofd' : '') +
          (tel(deze, 'laat') ? ' · ' + tel(deze, 'laat') + '× te laat' : '') +
          (tel(deze, 'geo') ? ' · ' + tel(deze, 'geo') + ' u geoorloofd' : '') +
          '</span>';

      var dagen = '';
      for (var d = 0; d < 5; d++) dagen += dagCel(l, dm, HUIDIG, d);

      return '<article class="kaart' + (rustig ? ' rustig' : '') + '">' +
        '<header class="kaart-kop">' +
          '<span class="wie"><span class="naam">' + esc(l.naam) + '</span>' +
            '<span class="sub">' + esc(sub) + '</span></span>' + cijfers +
        '</header>' +
        '<div class="week">' + dagen + '</div>' +
        '<div class="kaart-voet">' +
          '<span class="eerder"><span class="eerder-kop">weken ervoor</span>' + streepjes(l) + '</span>' +
          '<span class="totaal">' + l.ong + ' u ongeoorloofd · ' + l.laat + '× te laat in ' +
            WEKEN.length + ' weken</span>' +
        '</div>' +
        logboekBlok(l) +
      '</article>';
    }).join('') || '<div class="leeg">Nog geen leerlingen opgehaald.</div>';
  }

  function dagenVan(l) { return perDatum(l); }

  /* ── Tooltip ────────────────────────────────────────────────────────────── */
  var tip = document.getElementById('tip'), huidig = null;
  document.addEventListener('mousemove', function (ev) {
    var t = ev.target.closest('[data-tip]');
    if (!t) { if (huidig) { tip.classList.remove('on'); huidig = null; } return; }
    if (t !== huidig) { huidig = t; tip.textContent = t.dataset.tip; tip.classList.add('on'); }
    var r = t.getBoundingClientRect();
    tip.style.left = Math.min(window.innerWidth - 12, Math.max(12, r.left + r.width / 2)) + 'px';
    tip.style.top = (r.top - 7) + 'px';
  });

  tekenKop();
  tekenKaarten();
})();

File: coordinator.py
"""
coordinator.py — weekbeeld voor een coördinator met een eigen lijst leerlingen.

Zelfde data als het teamleider-dashboard (zie dashboard.verwerk), maar een
andere vraag: niet "wie springt eruit in deze afdeling", maar "hoe ging deze
week bij mijn twintig". Daarom staat de huidige week uitgeschreven per lesdag
en staan de weken ervoor als streepjes.

De opmaak leunt op template.html: daar staan de kleuren en de basisstijlen, en
die willen we niet twee keer onderhouden.
"""

import re
import json
from pathlib import Path

import dashboard

HIER = Path(__file__).parent


def _basis_stijl():
    """Alles tot en met </head> uit het teamleider-sjabloon: tokens en basis."""
    sjabloon = (HIER / 'template.html').read_text(encoding='utf-8')
    return sjabloon[:sjabloon.index('</head>')]


EIGEN_STIJL = """
<style>
.kop-rij { display: flex; align-items: flex-start; justify-content: space-between;
  gap: 16px; flex-wrap: wrap; }
#weeklabel { font-weight: 600; color: var(--ink); }

.kaarten { display: flex; flex-direction: column; gap: 10px; }
.kaart { background: var(--card); border: 1px solid var(--lijn); border-radius: var(--radius);
  padding: 12px 14px; box-shadow: var(--shadow); }
.kaart.rustig { opacity: .72; }
.kaart-kop { display: flex; align-items: baseline; justify-content: space-between;
  gap: 12px; flex-wrap: wrap; margin-bottom: 9px; }
.kaart .wie { padding: 0; }
.kaart .naam { display: block; font-weight: 600; font-size: .92rem; }
.kaart .sub { display: block; font-size: .74rem; color: var(--muted); margin-top: 1px; }
.wcijfers { font-size: .78rem; color: var(--ink-2); white-space: nowrap; }
.wcijfers b { color: var(--ink); }
.wrust { font-size: .78rem; color: var(--muted); font-style: italic; }

/* De week zelf: vijf lesdagen naast elkaar */
.week { display: grid; grid-template-columns: repeat(5, 1fr); gap: 6px; }
.wdag { border: 1px solid var(--lijn); border-radius: 8px; padding: 6px 7px 8px;
  background: var(--card-sunk); min-height: 58px; }
.wdag.ong  { border-color: color-mix(in srgb, var(--s-ongeoorloofd) 55%, transparent);
             background: color-mix(in srgb, var(--s-ongeoorloofd) 8%, var(--card)); }
.wdag.laat { border-color: color-mix(in srgb, var(--s-telaat) 55%, transparent);
             background: color-mix(in srgb, var(--s-telaat) 8%, var(--card)); }
.wdag.geo  { border-color: color-mix(in srgb, var(--s-geoorloofd) 45%, transparent);
             background: color-mix(in srgb, var(--s-geoorloofd) 7%, var(--card)); }
.wdag-kop { display: block; font-size: .68rem; font-weight: 600; color: var(--muted);
  text-transform: uppercase; letter-spacing: .05em; margin-bottom: 4px; }
.wdag-body { display: flex; flex-wrap: wrap; gap: 3px; }
.wchip { font-size: .7rem; font-weight: 700; padding: 1px 5px; border-radius: 5px;
  background: var(--card); box-shadow: inset 0 0 0 1px var(--lijn); color: var(--ink-2); }
.wchip.ong  { background: color-mix(in srgb, var(--s-ongeoorloofd) 22%, var(--card));
              color: var(--s-ongeoorloofd); box-shadow: none; }
.wchip.laat { background: color-mix(in srgb, var(--s-telaat) 24%, var(--card));
              color: #8a6100; box-shadow: none; }
.wchip.geo  { background: color-mix(in srgb, var(--s-geoorloofd) 18%, var(--card));
              color: var(--s-geoorloofd); box-shadow: none; }
.wleeg { color: var(--line-soft); }

.kaart-voet { display: flex; align-items: center; justify-content: space-between;
  gap: 12px; flex-wrap: wrap; margin-top: 9px; }
.eerder { display: inline-flex; align-items: center; gap: 6px; }
.eerder-kop { font-size: .68rem; color: var(--muted); text-transform: uppercase;
  letter-spacing: .05em; }
.totaal { font-size: .72rem; color: var(--muted); font-variant-numeric: tabular-nums; }
.kaart .logboek { margin: 8px 0 0; }
@media (max-width: 760px) { .week { grid-template-columns: repeat(3, 1fr); } }
</style>
</head>
<body>
"""


def bouw_html(payload, codes=None, config=None, mentoren=None,
              patroon=dashboard.MENTORGROEP_PATROON, met_logboek=True, bron=''):
    """Weekoverzicht als zelfstandig HTML-bestand."""
    codes = codes if codes is not None else dashboard.laad_codes()
    data, info = dashboard.verwerk(payload, codes, config, mentoren, patroon, met_logboek)

    periode = data['periode']
    lichaam = f"""
<div class="wrap">
  <div class="masthead kop-rij">
    <div>
      <h1>Mijn leerlingen</h1>
      <p>Hoe ging deze week, en wat liep er in de weken ervoor?</p>
    </div>
    <div class="meta">
      <b id="weeklabel"></b>
      {info['aantal_leerlingen']} leerlingen · {periode['label']}
    </div>
  </div>

  <div class="tiles" id="tiles"></div>

  <div class="kaarten" id="kaarten"></div>

  <footer>
    <b>Hoe je dit leest.</b> Bovenaan staat de leerling waar deze week het meest
    speelde. De vijf vakken zijn de lesdagen van deze week met de codes die er
    staan; de streepjes daaronder zijn de weken ervoor, één blokje per lesdag.
    Ziek en verlof staan er als context bij en tellen niet als ongeoorloofd.
    {bron}
  </footer>
</div>

<div id="tip" role="status" aria-live="polite"></div>

<script>window.DATA = {json.dumps(data, ensure_ascii=False).replace('<', chr(92) + 'u003c')};</script>
<script>
{(HIER / 'coordinator.js').read_text(encoding='utf-8')}
</script>
</body>
</html>
"""
    return _basis_stijl() + EIGEN_STIJL + lichaam, info


def lees_ids(tekst):
    """Leerlingnummers uit een vrij ingetypte lijst halen.

    Alles wat geen cijfer is geldt als scheidingsteken, zodat plakken uit
    Excel, een mail of een kommalijst allemaal werkt.
    """
    gezien, uit = set(), []
    for stuk in re.findall(r'\d+', tekst or ''):
        nummer = int(stuk)
        if nummer not in gezien:
            gezien.add(nummer)
            uit.append(nummer)
    return uit
